src/Deserialization.cpp

```diff
@@ -0,0 +1,1358 @@
+            buffer = it->second;
+        }
+        return ExternFuncArgument(buffer);
+    } else if (arg_type == ExternFuncArgument::ArgType::ExprArg) {
+        auto expr = deserialize_expr(extern_func_argument->expr_type(), extern_func_argument->expr());
+        return ExternFuncArgument(expr);
+    } else {
+        auto image_param_name = deserialize_string(extern_func_argument->image_param_name());
+        Parameter image_param;
+        if (auto it = non_serialized_parameters.find(image_param_name); it != non_serialized_parameters.end()) {
+            image_param = it->second;
+        } else if (auto it = parameters_in_pipeline.find(image_param_name); it != parameters_in_pipeline.end()) {
+            image_param = it->second;
+        }
+        return ExternFuncArgument(image_param);
+    }
+}
+
+Buffer<> Deserializer::deserialize_buffer(const Serialize::Buffer *buffer) {
+    assert(buffer != nullptr);
+    if (!buffer->defined()) {
+        return Buffer<>();
+    }
+    std::string name = deserialize_string(buffer->name());
+    auto type = deserialize_type(buffer->type());
+    int32_t dimensions = buffer->dimensions();
+    std::vector<halide_dimension_t> buffer_dimensions;
+    buffer_dimensions.reserve(dimensions);
+    for (int i = 0; i < dimensions; ++i) {
+        auto dim = buffer->dims()->Get(i);
+        halide_dimension_t hl_dim;
+        hl_dim.min = dim->min();
+        hl_dim.extent = dim->extent();
+        hl_dim.stride = dim->stride();
+        buffer_dimensions.push_back(hl_dim);
+    }
+    auto fake_buffer = Buffer<>(type, nullptr, dimensions, buffer_dimensions.data(), name + "_fake");
+    auto hl_buffer = Buffer<>::make_with_shape_of(fake_buffer, nullptr, nullptr, name);
+    memcpy(hl_buffer.data(), buffer->data()->data(), buffer->data()->size());
+    return hl_buffer;
+}
+
+std::map<std::string, FunctionPtr> Deserializer::deserialize_wrapper_refs(const flatbuffers::Vector<flatbuffers::Offset<Serialize::WrapperRef>> *wrappers) {
+    assert(wrappers != nullptr);
+    std::map<std::string, FunctionPtr> result;
+    for (const auto &wrapper : *wrappers) {
+        auto name = deserialize_string(wrapper->func_name());
+        int32_t func_index = wrapper->func_index();
+        FunctionPtr func_ptr;
+        if (auto it = this->reverse_function_mappings.find(func_index); it != this->reverse_function_mappings.end() && func_index != -1) {
+            func_ptr = it->second;
+        }
+        result[name] = func_ptr;
+    }
+    return result;
+}
+
+void Deserializer::build_reverse_function_mappings(const std::vector<Function> &functions) {
+    if (!this->reverse_function_mappings.empty()) {
+        this->reverse_function_mappings.clear();
+    }
+    int cnt = 0;
+    for (const auto &f : functions) {
+        this->reverse_function_mappings[cnt++] = f.get_contents();
+    }
+}
+
+Pipeline Deserializer::deserialize(const std::string &filename) {
```

Comment:
This would be a lot more flexible if you split it into two functions, with one taking `std::istream&` as an argument (note, *not* an `ifstream`), so that callers could use serialization to/from memory, or some other thing that doesn't map well to a POSIX file.